[FIX] point_of_sale: reduce memory usage when loading products

jorenvo opened the pull request:
During the search_read of product.product various fields are
prefetched without ever being used. For ~70,000 products disabling
prefetching completely reduces memory usage from ~2 GB to ~1 GB
without affecting load time.

opw-1913872

nim-odoo commented:
If it decreases the memory usage, it also probably increases the number of DB requests. I would also go for a read, if possible.

nim-odoo commented:
@nseinlet Note that the PoS performs a `search_read` call:
https://github.com/odoo/odoo/blob/d9b49c8870e116296a7e8f7896c8c694d87306ef/addons/point_of_sale/static/src/js/models.js#L565

Which does a `search` followed by a `read`:
https://github.com/odoo/odoo/blob/d9b49c8870e116296a7e8f7896c8c694d87306ef/odoo/models.py#L4496-L4532

So I would guess that the `read` still causes the prefetching of unnecessary fields.

jorenvo commented:
Indeed I came to the same conclusion yesterday. The `search_read` already does the `read` but it will still optimistically cache prefetchable fields. A good example of this is `description_pickingin`, it's used only in a view and a report. I did some deeper investigation and by the end of the `search_read` the cache holds the following fields:

```
product.product.pos_categ_id: 70531 records
product.product.to_weight: 70531 records
product.product.uom_id: 70531 records
product.product.description_sale: 70531 records
product.product.description: 70531 records
product.product.tracking: 70531 records
product.template.display_name: 70000 records
product.template.default_code: 70000 records
product.template.description: 70000 records
product.template.type: 70000 records
product.template.to_weight: 70000 records
product.template.create_uid: 70000 records
product.template.color: 70000 records
product.template.service_type: 70000 records
product.template.sequence: 70000 records
product.template.description_purchase: 70000 records
product.template.sale_delay: 70000 records
product.template.tracking: 70000 records
product.template.purchase_line_warn_msg: 70000 records
product.template.write_uid: 70000 records
product.template.volume: 70000 records
product.template.sale_line_warn: 70000 records
product.template.rental: 70000 records
product.template.description_sale: 70000 records
product.template.description_picking: 70000 records
product.template.purchase_method: 70000 records
product.template.sale_line_warn_msg: 70000 records
product.template.uom_id: 70000 records
product.template.responsible_id: 70000 records
product.template.pos_categ_id: 70000 records
product.template.weight: 70000 records
product.template.service_to_purchase: 70000 records
product.template.expense_policy: 70000 records
product.template.sale_ok: 70000 records
product.template.write_date: 70000 records
product.template.categ_id: 70000 records
product.template.description_pickingout: 70000 records
product.template.purchase_line_warn: 70000 records
product.template.name: 70000 records
product.template.message_main_attachment_id: 70000 records
product.template.uom_po_id: 70000 records
product.template.invoice_policy: 70000 records
product.template.create_date: 70000 records
product.template.purchase_ok: 70000 records
product.template.list_price: 70000 records
product.template.description_pickingin: 70000 records
product.template.available_in_pos: 70000 records
product.template.company_id: 70000 records
product.template.active: 70000 records
product.category.display_name: 39 records
product.category.complete_name: 39 records
product.category.removal_strategy_id: 39 records
product.category.parent_path: 39 records
product.category.write_uid: 39 records
product.category.write_date: 39 records
product.category.create_date: 39 records
product.category.parent_id: 39 records
product.category.create_uid: 39 records
product.category.name: 39 records
product.product.taxes_id: 70531 records
product.template.taxes_id: 70531 records
uom.uom.display_name: 451 records
uom.uom.name: 451 records
uom.uom.create_uid: 451 records
uom.uom.active: 451 records
uom.uom.rounding: 451 records
uom.uom.category_id: 451 records
uom.uom.write_date: 451 records
uom.uom.measure_type: 451 records
uom.uom.create_date: 451 records
uom.uom.uom_type: 451 records
uom.uom.factor: 451 records
uom.uom.write_uid: 451 records
product.product.lst_price: 70531 records
product.product.price_extra: 70531 records
product.product.product_template_attribute_value_ids: 70531 records
product.product.attribute_value_ids: 70531 records
product.product.standard_price: 70000 records
res.users.company_id: 1 records
res.users.write_date: 1 records
res.users.sale_team_id: 1 records
res.users.pos_security_pin: 1 records
res.users.notification_type: 1 records
res.users.partner_id: 1 records
res.users.create_uid: 1 records
res.users.alias_id: 1 records
res.users.create_date: 1 records
res.users.write_uid: 1 records
res.users.active: 1 records
res.users.odoobot_state: 1 records
res.users.login: 1 records
res.users.share: 1 records
res.users.signature: 1 records
res.users.action_id: 1 records
product.template.standard_price: 0 records
product.product.stock_value: 0 records
product.product.qty_at_date: 0 records
product.product.stock_fifo_real_time_aml_ids: 0 records
product.product.stock_fifo_manual_move_ids: 0 records
pos.category.display_name: 37 records
pos.category.name: 37 records
pos.category.create_date: 37 records
pos.category.write_uid: 37 records
pos.category.write_date: 37 records
pos.category.create_uid: 37 records
pos.category.sequence: 37 records
pos.category.parent_id: 37 records
total_cached_fields: 4285141
```

I think about half of those are not used by the POS in any way, which explains the 50% memory reduction. I agree disabling prefetching fully is not great for the DB, because it will result in a lot more queries.

Is there some way to limit prefetching to a subset of fields? I looked around but I can't find anything like that.

nim-odoo commented:
So there could be other workarounds:

- The prefetching of `read` is activated when some non-stored computed fields require fields which are not in the cache. It could be possible to run first the `search_read` with `prefetch_fields=False`, check what are the extra fields fetched, then add explicitly these fields in `model.js`. But that doesn't seem very robust: if one needs a new non-stored computed field, you have to do the job all over again.

- Use an offset in the `search_read` to fetch the information in several batches. There are several possibilities to do it, but I think the one which would have the least impact in the code would be to:
1. Execute a `search_count` first to get the total number of records
2. Execute the necessary `search_read` calls (not in parallel, though)

This way, you would probably avoid a MemoryError as well as a time out, although the total loading time might be slightly increased. Indeed, you would have several "light" calls instead of a single "heavy" call. This could also probably be applied only to `product.product` and `res.partner` models.

cc @rim-odoo

jorenvo commented:
Your second approach is probably the best thing to do regardless of the possible prefetching improvements. Even if the prefetching was optimized you'd end up with exact same issue if you have 140,000 products instead of 70,000. Loading in batches scales better.

I'm closing this because I've implemented it in v11: https://github.com/odoo/odoo/pull/29489. I rebased on v11 because these MemoryErrors are made possible by the following 2 v11 fixes by RCO if I remember correctly:

d39d583b213af02bad12721d9c610def6cb09161
70ed882088401c093e6a1180746b5637fe96da32

before these it would time out before running out of memory.
